code/mob/living/critter/ai/sawflyai.dm

```diff
@@ -119,24 +122,23 @@
 	..()
 
 /datum/aiTask/sequence/goalbased/sawfly_chase_n_stab/get_targets()
-	. = list()
+	. = list() // don't ask why the list is simply a period
 	var/mob/living/critter/robotic/sawfly/owncritter = holder.owner
-	for (var/mob/living/C in view(owncritter,max_dist))
-		if(C == owncritter) continue
+	for (var/mob/living/C in viewers(max_dist, owncritter))
 		if (C.health < -50 || !isalive(C)) continue
-		if (C.job == "Security Officer" || C.job == "Head of Security")
-			. = list(C) //found a secoff, just return that
-			return
 		if(istype(C, /mob/living/critter/robotic/sawfly)) continue
-		if (C in owncritter.friends) continue
-		if (istraitor(C) || isnukeop(C) || isspythief(C)) // frens :)
-			boutput(C, "<span class='alert'> The [owncritter]'s IFF system silently flags you as an ally! </span>")
-			owncritter.friends += C
-			continue
+		if(C.mind?.special_role)
+			if (istraitor(C) || isnukeop(C) || isspythief(C) || isnukeopgunbot(C)) // frens :)
+				if (!(C in owncritter.friends))
+					boutput(C, "<span class='alert'> [owncritter]'s IFF system silently flags you as an ally! </span>")
+					owncritter.friends += C
+				continue
 		. += C //you passed all the checks it, now you get added to the list for consideration
+		if(C.job in list( "Head of Security", "Security Officer", "Nanotrasen Security Consultant")) //hopefully this is cheaper than the OR chain I had before
+			. = list(C) //go get em, tiger
+			return
 	. = get_path_to(holder.owner, ., max_dist*2, 1) //calculate paths to the target, any unreachable targets will be discarded
```

Comment:
same comments here. Don't bother with refs or use weak refs. Return on targets early, here you are using `get_path_to` which is exceptionally expensive so probably just return on the very first target.